### PR TITLE
VBM-76: Sign Up added

### PR DESCRIPTION
**Pull request by VladimirBludov**

### CODE REVIEW

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 1]
```diff
@@ -1,21 +1,158 @@
 import * as Joi from 'joi';
-import { UserValidationMessage, UserSignUpDto } from '@vse-bude/shared';
+import { UserSignUpDto } from '@vse-bude/shared';
+import i18next from 'i18next';
+
+const emailValidator = (value: string, helpers: Joi.CustomHelpers) => {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/validation-schemas/user/sign-up.ts:5]: I'd suggest extracting all these validation helpers either to `helpers/validation` or to `validation-schemas/helpers` as you might need some of those to be used in other schemas in future

File: packages/mobile/src/localization/languages/en/screens.ts [hunk 2]
```diff
@@ -47,4 +47,26 @@ export const screens = {
...
+  errors: {
+    EMPTY_EMAIL: 'Fill the email please',
+    EMPTY_FIRST_NAME: 'Fill the first name please!',
+    EMPTY_LAST_NAME: 'Fill the last name please!”',
+    EMPTY_PHONE: 'Fill the phone number please!',
+    EMPTY_PASSWORD: 'Fill the password please!',
+    LATIN: 'You can use only Latin letters',
+    INVALID_EMAIL: 'The invalid email format',
+    SPACES: 'Spaces are not allowed',
+    LENGTH_8: 'Password length must be at least 8 characters',
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/localization/languages/en/screens.ts:59]: corrected

File: packages/mobile/src/localization/languages/en/screens.ts [hunk 3]
```diff
@@ -47,4 +47,26 @@ export const screens = {
...
+    LENGTH_16: 'Password length must not exceed 16 characters',
+    UP_LOW_NUM:
+      'Password should include at least one uppercase, one lowercase and one number',
+    INVALID_PASSWORD: 'The invalid password format',
+    REPEAT_PASSWORD_INVALID: 'Passwords are not the same',
+    FIRST_NAME_LENGTH_1: 'First name must be at least 1 character',
+    FIRST_NAME_LENGTH_40: 'First name must be not more than 40 characters',
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/localization/languages/en/screens.ts:66]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 4]
```diff
@@ -1,21 +1,158 @@
...
+  if (value.match(/[а-яА-Я]/)) {
+    return helpers.message({ custom: i18next.t('errors.LATIN') });
+  }
+
+  if (
+    !value.match(
+      /([0-9a-zA-Z!#$%&'*+-/=?^_`{|}~.]{3,64})@([0-9a-zA-Z]+)\.([0-9a-zA-Z]{2,64})/,
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:12]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 5]
```diff
@@ -1,21 +1,158 @@
...
+    return helpers.message({ custom: i18next.t('errors.INVALID_EMAIL') });
+  }
+
+  return value;
+};
+
+const passwordValidator = (value: string, helpers: Joi.CustomHelpers) => {
+  const isUppercase = value.match(/[A-Z]/);
+  const isLowercases = value.match(/[a-z]/);
+  const isNumber = value.match(/[0-9]/);
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:24]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 6]
```diff
@@ -1,21 +1,158 @@
...
+
+  if (value.match(/[а-яА-Я]/)) {
+    return helpers.message({ custom: i18next.t('errors.LATIN') });
+  }
+
+  if (value.match(/\s/)) {
+    return helpers.message({ custom: i18next.t('errors.SPACES') });
+  }
+
+  if (value.length < 8) {
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:34]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 7]
```diff
@@ -1,21 +1,158 @@
...
+  }
+
+  if (value.match(/^-|-$/)) {
+    return helpers.message({ custom: i18next.t('errors.FIRST_NAME_INVALID') });
+  }
+
+  return value;
+};
+
+const lastNameValidator = (value: string, helpers: Joi.CustomHelpers) => {
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:85]: we using different constants length for first/last name, so different validators
corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 8]
```diff
@@ -1,21 +1,158 @@
...
+  }
+
+  return value;
+};
 
 const signUp = Joi.object<UserSignUpDto>({
   email: Joi.string()
     .trim()
+    .custom(emailValidator, 'custom validation')
     .email({ tlds: { allow: false } })
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:115]: we extend joi email validations with our own
corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 9]
```diff
@@ -1,21 +1,158 @@
...
+    }),
+  lastName: Joi.string()
+    .trim()
+    .custom(lastNameValidator, 'custom validation')
+    .required()
+    .messages({
+      'string.empty': i18next.t('errors.EMPTY_LAST_NAME'),
+    }),
+  phone: Joi.string()
+    .pattern(/^\+\d{8,15}$/)
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:135]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (AleksandrTermenzhy) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:85]: In that case you may want to create a curried function, e.g.:
```
const createJoiCustomStringValidator = (minLength: number, maxLength: number) => (value: string, helpers: Joi.CustomHelpers) => {
    // some logic goes here
}
// ...
const lastNameValidator = createJoiCustomStringValidator(5,5);
const firstNameValidator = createJoiCustomStringValidator(3,20);
```

Also you may add fieldName if needed

[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:85]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts [hunk 10]
```diff
@@ -1,21 +1,158 @@
...
+      'string.empty': i18next.t('errors.EMPTY_PHONE'),
+    }),
+  password: Joi.string()
+    .custom(passwordValidator)
+    .trim()
+    .required()
+    .messages({
+      'string.empty': i18next.t('errors.EMPTY_PASSWORD'),
+    }),
+  repeatPassword: Joi.any()
```
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:149]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts (hunk 1 above)
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts:5]: corrected

File: packages/mobile/src/validation-schemas/user/sign-up.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (VladimirBludov) [reply packages/mobile/src/validation-schemas/user/sign-up.ts]: corrected

**Review by AleksandrTermenzhy — APPROVED** [review]

**Pull request merged**